### PR TITLE
build: introduce lerna changed command to perform some optimized build

### PR DESCRIPTION
**Pull request by antleblanc**
| Question         | Answer
| ---------------- | ---
| Branch?          | `develop`
| Bug fix?         | no
| New feature?     | yes
| Breaking change? | yes
| Tickets          | MANAGER-5923
| License          | BSD 3-Clause

### :construction_worker: Build

a85b2b1  - build: introduce lerna changed command to perform some optimized build

### :link: Related

- https://github.com/lerna/lerna/tree/master/commands/changed

### :bookmark: TODO

- [x] Bring back option `num-workers`.
- [x] Bring back option `dry-run`?.
- [x] Validate bump of `commander` across existing CLIs.
- [x] Validate changes against pipelines.

Signed-off-by: Antoine Leblanc <antoine.leblanc@ovhcloud.com>

### CODE REVIEW

File: scripts/worker/build_module.js [hunk 1]
```diff
@@ -1,11 +1,14 @@
 const { parentPort, workerData } = require('worker_threads');
 const execa = require('execa');
+const yn = require('yn');
 
 execa
-  .command(`cd ${workerData.location} && npm run build --if-present`, {
+  // .command(`lerna run --scope ${workerData.name} build`, {
+  .command(`cd ${workerData.location} && pwd`, {
```
Comment (antleblanc) [inline scripts/worker/build_module.js:7]: For testing purpose only! Will be removed!

**Comment by ovh-cds** [comment]
CDS Report build#13998.0 ✘  
* 
  * Build ✘

**Comment by ovh-cds** [comment]
CDS Report test-manager_2#13998.0 ✘  
* Stage 1
  * e2e -   
  * lint ✔ 
  * SemVer ✘

**Comment by cbourgois** [comment]
Build `workspace` parameter added in this PR : https://github.com/ovh/manager/pull/3262
To allow to build packages for a specific workspace.

**Comment by ovh-cds** [comment]
CDS Report test#16861.0 ✘  
* Stage 1
  * lint ✔ 
  * SemVer ✘

**Comment by cbourgois** [comment]
Rebased on develop, include `workspace` parameter and restore `numWorkers` and `dryRun`

File: package.json [hunk 2]
```diff
@@ -87,7 +90,9 @@
     "stylelint-scss": "^3.18.0",
     "testcafe": "^1.8.4",
     "testcafe-angular-selectors": "^0.4.1",
-    "testcafe-reporter-html": "^1.4.6"
+    "testcafe-reporter-html": "^1.4.6",
+    "vuepress": "^1.2.0",
```
Comment (antleblanc) [inline package.json:94]:
```suggestion
```

Comment (cbourgois) [reply package.json]: fixed.

File: package.json [hunk 3]
```diff
@@ -87,7 +90,8 @@
     "stylelint-scss": "^3.18.0",
     "testcafe": "^1.8.4",
     "testcafe-angular-selectors": "^0.4.1",
-    "testcafe-reporter-html": "^1.4.6"
+    "testcafe-reporter-html": "^1.4.6",
+    "yn": "^4.0.0"
```
Comment (cbourgois) [inline package.json:94]: Seems not required.

Comment (antleblanc) [reply package.json]: You're right! `yn` can be safely removed from the root `package.json` file.

Only used here:
```sh
$ git grep "import yn" -- "*.ts"
packages/manager/tools/webpack-dev-server/src/config.ts:import yn from 'yn';
```

**Comment by ovh-cds** [comment]
CDS Report build#16894.0 ✘  
* 
  * Build ✘

File: scripts/build.js [hunk 4]
```diff
@@ -167,15 +203,23 @@ program
     '-w, --workspace [workspace]',
     'Scope build to a specific workspace, its packages and their dependencies',
   )
+  .option('--verbose', 'output extra debugging')
   .action(async () => {
+    process.env.VERBOSE = program.verbose;
```
Comment (marie-j) [inline scripts/build.js:208]: why does it need to be a `env` variable ?

File: yarn.lock [hunk 5]
```diff
@@ -7171,16 +7189,16 @@ commander@5.1.0:
   resolved "https://registry.yarnpkg.com/commander/-/commander-5.1.0.tgz#46abbd1652f8e059bddaef99bbdcb2ad9cf179ae"
   integrity sha512-P0CysNDQ7rtVw4QIQtm+MRxV66vKFSvlsQvGYXZWR3qFU0jlMKHZZZgw8e+8DSah4UDKMqnknRDQz+xuQXQ/Zg==
 
-commander@^4.0.0:
-  version "4.0.1"
-  resolved "https://registry.yarnpkg.com/commander/-/commander-4.0.1.tgz#b67622721785993182e807f4883633e6401ba53c"
-  integrity sha512-IPF4ouhCP+qdlcmCedhxX4xiGBPyigb8v5NeUp+0LyhwLgxMqyp3S0vl7TAPfS/hiP7FC3caI/PB9lTmP8r1NA==
-
-commander@^4.1.1:
+commander@^4.0.0, commander@^4.1.1:
```
Comment (marie-j) [inline yarn.lock:7192]: needed for another deps ?

Comment (cbourgois) [reply yarn.lock:7192]: Required by some vendors (`html-minifier-terser`, `gh-pages`...)

File: scripts/build.js (hunk 4 above)
Comment (antleblanc) [reply scripts/build.js:208]: Environment variable need to be passed to the Worker Thread via [`SHARE_ENV`](https://nodejs.org/api/worker_threads.html#worker_threads_worker_share_env).

Used here: https://github.com/ovh/manager/pull/3328/files#diff-a992dd64a8c626fedc95d613f43602b91f6dc6339893afaf405dc05fa13c9ceaR150

It helps us to output extra logs if needed.

**Comment by cbourgois** [comment]
Rebased on `develop`

**Review by cbourgois — APPROVED** [review]

**Review by marie-j — APPROVED** [review]

**Review by JeremyDec — DISMISSED** [review]

**Review by cbourgois — APPROVED** [review]

File: scripts/build.js [hunk 6]
```diff
@@ -167,15 +253,24 @@ program
...
+
     let packagesInfos = [];
 
     if (program.workspace) {
       packagesInfos = await getPackagesByWorkspace(
         path.resolve(program.workspace),
       );
     } else if (program.package) {
-      packagesInfos = await getPackagesByPackageName(program.package);
+      // TODO: use variadic option instead of passing an array.
```
Comment (marie-j) [inline scripts/build.js:270]: In another PR ?

**Review by marie-j — APPROVED** [review]

File: scripts/build.js (hunk 6 above)
Comment (cbourgois) [reply scripts/build.js:270]: Yes, in future PR

**Pull request merged**